---
title: "Pseudotime Biodomain Networks & KDs"
author: "Greg Cary"
date: "`r Sys.Date()`"
output: 
  html_document: 
    fig_width: 10
    fig_height: 10
  html_notebook:
    code_folding: hide
    theme: spacelab
    css: style.css
    toc: true
    toc_float: true
---

```{r setup}

library(synapser)
library(igraph)
library(tidyverse)
source('/projects/carter-lab/caryg/biodom_tally.R')
# source(paste0(here::here(), '/scripts/Mergeomics_Version_1.99.0.R'))

theme_set(theme_bw())

```

Grab relevant data from synapse
```{r message=FALSE, warning=FALSE}

synLogin()

# target risk scores
scores <- read_csv(synTableQuery('select * from syn25575156', includeRowIdAndRowVersion = F)$filepath)

# biological domain annotations
biodom <- full_join(
  # biodomains
  readRDS(synGet('syn25428992')$path),
  # domain labels
  read_csv(synGet('syn26856828')$path),
  by = c('Biodomain'='domain')
) %>%
  mutate(Biodomain = case_when(Biodomain == 'none' ~ NA_character_, T ~ Biodomain))

# biodomain enrichment results based on risk scores
bd_enr_res = read_csv(synapser::synGet('syn45824995')$path) %>% 
  # filter(Biodomain != 'none') %>% 
  mutate(leadingEdge_genes = str_split(leadingEdge_genes, '\\|'))

# ROSMAP txomic PT
rna.f.de <- read_csv( synapser::synGet('syn39989123')$path, col_types = cols() ) %>% rename(gene = gene_names)
rna.f.enr <- read_tsv( synapser::synGet('syn47728345')$path, col_types = cols() )
rna.m.de <- read_csv( synapser::synGet('syn39990047')$path, col_types = cols() ) %>% rename(gene = gene_names)
rna.m.enr <- read_tsv( synapser::synGet('syn47728065')$path, col_types = cols() )

# ROSMAP proteomic PT
prot.f.de <- read_csv( synapser::synGet('syn40616521')$path, col_types = cols() ) %>% rename(gene = gene_short_name)
prot.f.enr <- read_csv( synapser::synGet('syn50881293')$path, col_types = cols() )
prot.m.de <- read_csv( synapser::synGet('syn40621972')$path, col_types = cols() ) %>% rename(gene = gene_short_name)
prot.m.enr <- read_csv( synapser::synGet('syn50881295')$path, col_types = cols() )


```

# Number of leading edge genes per state vs significance level for each dataset

```{r}

enr <- prot.f %>% 
  filter(!is.na(Biodomain)) %>% 
  mutate(leadingEdge = `enr2$leadingEdge`  %>% str_remove_all(.,'^c|[:punct:]') %>% str_split(' ')  )

sig_explore = map_dfr(
  c(1e-2, 1e-3, 1e-4, 1e-5, 1e-6, 1e-7),
  ~ enr %>% 
      filter( padj < .x
              # , abs(NES) > 2 
              ) %>% 
      group_by( state ) %>%  
      summarise( 
        n_sig_bd = length(unique(Biodomain)),
        n_sig_term = length(unique(pathway)),
        n_leadingEdge_genes = length(unique(unlist(leadingEdge)))) %>% 
      arrange(state) %>% 
      mutate( sig_level = .x)
)

ggplot(sig_explore, aes(n_leadingEdge_genes, as.factor(state) )) +
  # geom_point() +
  geom_bar(stat = 'identity') +
  scale_y_discrete( limits = rev )+
  facet_wrap(~sig_level) +
  ggtitle('F proteomics pseudotime'
          # , subtitle = 'only terms with abs(NES) > 2'
          )

```

```{r}

enr <- prot.m %>% 
  filter(!is.na(Biodomain)) %>% 
  mutate(leadingEdge = `enr2$leadingEdge` %>% str_remove_all(.,'^c|[:punct:]') %>% str_split(' ')  )

sig_explore = map_dfr(
  c(1e-2, 1e-3, 1e-4, 1e-5, 1e-6, 1e-7),
  ~ enr %>% 
      filter( padj < .x ) %>% 
      group_by( state ) %>% 
      summarise( 
        n_sig_bd = length(unique(Biodomain)),
        n_sig_term = length(unique(pathway)),
        n_leadingEdge_genes = length(unique(unlist(leadingEdge)))) %>% 
      arrange(state) %>% 
      mutate( sig_level = .x)
)

ggplot(sig_explore, aes(n_leadingEdge_genes, as.factor(state) )) +
  # geom_point() +
  geom_bar(stat = 'identity') +
  scale_y_discrete( limits = rev )+
  facet_wrap(~sig_level) +
  ggtitle('M proteomics pseudotime')

```

```{r}

enr <- r.rna.f %>% 
  filter(!is.na(Biodomain)) %>% 
  mutate(leadingEdge = `enr2$leadingEdge` %>% str_remove_all(.,'^c|[:punct:]') %>% str_split(' ') )

sig_explore = map_dfr(
  c(1e-2, 1e-4, 1e-6, 1e-8, 1e-10, 1e-12),
  ~ enr %>% 
      filter( padj < .x
              # , abs(NES) > 2.5
              ) %>% 
      group_by( state ) %>% 
      summarise( 
        n_sig_bd = length(unique(Biodomain)),
        n_sig_term = length(unique(pathway)),
        n_leadingEdge_genes = length(unique(unlist(leadingEdge)))) %>% 
      arrange(state) %>% 
      mutate( sig_level = .x)
)

ggplot(sig_explore, aes(n_leadingEdge_genes, as.factor(state) )) +
  # geom_point() +
  geom_bar(stat = 'identity') +
  scale_y_discrete( limits = rev )+
  facet_wrap(~sig_level) +
  ggtitle('F transcriptomics pseudotime (AMP-AD v2.0)'
          # , subtitle = 'only terms with abs(NES) > 2.5'
          )

```

```{r}

enr <- r.rna.m %>% 
  filter(!is.na(Biodomain)) %>% 
  mutate(leadingEdge = `enr2$leadingEdge` %>% str_remove_all(.,'^c|[:punct:]') %>% str_split(' ')  )

sig_explore = map_dfr(
  c(1e-2, 1e-4, 1e-6, 1e-8, 1e-10, 1e-12),
  ~ enr %>% 
      filter( padj < .x
              # , abs(NES) > 2.5 
              ) %>% 
      group_by( state ) %>% 
      summarise( 
        n_sig_bd = length(unique(Biodomain)),
        n_sig_term = length(unique(pathway)),
        n_leadingEdge_genes = length(unique(unlist(leadingEdge)))) %>% 
      arrange(state) %>% 
      mutate( sig_level = .x)
)

ggplot(sig_explore, aes(n_leadingEdge_genes, as.factor(state) )) +
  # geom_point() +
  geom_bar(stat = 'identity') +
  scale_y_discrete( limits = rev )+
  facet_wrap(~sig_level) +
  ggtitle('M transcriptomics pseudotime (AMP-AD v2.0)'
          # , subtitle = 'only terms with abs(NES) > 2.5'
          )

```

# Pseudotime states
## prot
```{r}
prot <- bind_rows(prot.f.de %>% mutate(s = 'f'), prot.m.de %>% mutate(s = 'm')) %>% 
    pivot_wider(id_cols = c(gene_names, gene, state), 
                values_from = c(pvalue, effect), names_from = s) %>% 
    rowwise() %>% 
    mutate(sig = case_when( 
        (pvalue_f <= 0.01 & pvalue_m > 0.01) ~ 'f_only',
        (pvalue_m <= 0.01 & pvalue_f > 0.01) ~ 'm_only',
        (pvalue_f > 0.01 & pvalue_m > 0.01) ~ 'neither',
        (pvalue_f <= 0.01 & pvalue_m <= 0.01) ~ 'both'),
        mn = mean(-log10(pvalue_f), -log10(pvalue_m), na.rm = T))
```

```{r fig.width=7, fig.height=5}
prot %>% 
  filter(sig != 'neither') %>% 
  ggplot(aes(effect_f, effect_m))+
  geom_hline(yintercept = 0, lwd =.1)+geom_vline(xintercept = 0, lwd =.1)+
  geom_point(aes(color = sig, size = mn), alpha = 0.3)+
  scale_size_continuous('mean -log10 pval', range = c(.1,2))+
  theme(legend.position = 'bottom')+
  facet_grid(sig~state)
```

```{r fig.width=9, fig.height=3}

cowplot::plot_grid(
  prot %>% 
    filter(sig != 'neither') %>% 
    group_by(state, sig) %>% 
    summarise(n_gene = length(unique(gene))) %>% 
    ggplot(aes(state, n_gene))+
    geom_col()+
    labs(subtitle = 'any effect direction')+
    facet_wrap(~sig),
  prot %>% 
    filter(sig == 'both', effect_f/effect_m > 0) %>% 
    group_by(state, sig) %>% 
    summarise(n_gene = length(unique(gene))) %>% 
    ggplot(aes(state, n_gene))+
    geom_col()+
    labs(subtitle = 'same effect direction')+
    facet_wrap(~sig)
  , rel_widths = c(1,.4)
)

```

```{r fig.width=9, fig.height=3}

tmp = prot %>% 
  filter (sig == 'both', effect_f/effect_m > 0 ) 

olap = crossing(st1 = unique(tmp$state), st2 = unique(tmp$state)) %>% 
  rowwise() %>% 
  mutate(
    olap = length(intersect(
      tmp %>% filter(state == st1) %>% pull(gene),
      tmp %>% filter(state == st2) %>% pull(gene) )),
    pct = olap / length(tmp %>% filter(state == st1) %>% pull(gene))
  ) %>% ungroup() %>% 
  mutate(st1 = factor(st1, levels = rev(c('1','2','3','4','5','6','7')) ),
         st2 = factor(st2, levels = c('1','2','3','4','5','6','7')))

```

```{r}
ggplot(olap, aes(x = st2, y = st1)) +
  geom_tile(aes(fill=pct), alpha = .8) + theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
  viridis::scale_fill_viridis("% Total", option = "C", direction = -1)+
  geom_text(aes(label = paste0(olap,'\n',signif(100*pct, 3), '%') ), size = 3) +
  labs(x='',y='', title = 'Proteomics Pseudotime State DE Genes') 
```

```{r fig.width=9, fig.height=3}

tmp = prot %>% 
  filter (sig == 'both', effect_f/effect_m > 0 ) %>% 
  mutate( state = case_when(
    state %in% c(2,3,4) ~ 'early', 
    state %in% c(5,6) ~ 'mid',
    state %in% c(7) ~ 'late',
    T ~ as.character(state)) )

olap = crossing(st1 = unique(tmp$state), st2 = unique(tmp$state)) %>% 
  rowwise() %>% 
  mutate(
    olap = length(intersect(
      tmp %>% filter(state == st1) %>% pull(gene),
      tmp %>% filter(state == st2) %>% pull(gene) )),
    pct = olap / length(tmp %>% filter(state == st1) %>% pull(gene))
  ) %>% ungroup() %>% 
  # mutate(st1 = factor(st1, levels = rev(c('early','5','6','7')) ),
  #        st2 = factor(st2, levels = c('early','5','6','7')))
    mutate(st1 = factor(st1, levels = rev(c('early','mid','late')) ),
         st2 = factor(st2, levels = c('early','mid','late')))

```

```{r}
ggplot(olap, aes(x = st2, y = st1)) +
  geom_tile(aes(fill=pct), alpha = .8) + theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
  viridis::scale_fill_viridis("% Total", option = "C", direction = -1)+
  geom_text(aes(label = paste0(olap,'\n',signif(100*pct, 3), '%') ), size = 3) +
  labs(x='',y='', title = 'Proteomics Pseudotime State DE Genes') 
```

## rna
```{r}
rna <- bind_rows(rna.f.de %>% mutate(s = 'f'), 
                 bind_rows(rna.m.de, rna.m.de %>% filter(state == 6) %>% mutate(state = 7)) %>% 
                   mutate(s = 'm')) %>% 
    pivot_wider(id_cols = c(gene, state), 
                values_from = c(pvalue, effect), 
                names_from = s) %>% 
    rowwise() %>% 
    mutate(sig = case_when( 
        (pvalue_f <= 0.001 & pvalue_m > 0.001) ~ 'f_only',
        (pvalue_m <= 0.001 & pvalue_f > 0.001) ~ 'm_only',
        (pvalue_f > 0.001 & pvalue_m > 0.001) ~ 'neither',
        (pvalue_f <= 0.001 & pvalue_m <= 0.001) ~ 'both'),
        mn = mean(-log10(pvalue_f), -log10(pvalue_m), na.rm = T))
```

```{r fig.width=7, fig.height=5}
rna %>% 
  filter(sig != 'neither') %>% 
  ggplot(aes(effect_f, effect_m))+
  geom_hline(yintercept = 0, lwd =.1)+geom_vline(xintercept = 0, lwd =.1)+
  geom_point(aes(color = sig, size = mn), alpha = 0.3)+
  scale_size_continuous('mean -log10 pval', range = c(.1,2))+
  theme(legend.position = 'bottom')+
  facet_grid(sig~state)
```

```{r fig.width=9, fig.height=3}

cowplot::plot_grid(
  rna %>% 
    filter(sig != 'neither') %>% 
    group_by(state, sig) %>% 
    summarise(n_gene = length(unique(gene))) %>% 
    ggplot(aes(state, n_gene))+
    geom_col()+
    labs(subtitle = 'any effect direction')+
    facet_wrap(~sig),
  rna %>% 
    filter(sig == 'both', effect_f/effect_m > 0) %>% 
    group_by(state, sig) %>% 
    summarise(n_gene = length(unique(gene))) %>% 
    ggplot(aes(state, n_gene))+
    geom_col()+
    labs(subtitle = 'same effect direction')+
    facet_wrap(~sig)
  , rel_widths = c(1,.4)
)

```

```{r fig.width=9, fig.height=3}

tmp = rna %>% 
  filter (sig == 'both', effect_f/effect_m > 0 ) %>% 
  mutate( state = case_when(
    state %in% c(2,3) ~ 'early', 
    state %in% c(4,5) ~ 'mid', 
    T ~ as.character(state)) )


olap = crossing(st1 = unique(tmp$state), st2 = unique(tmp$state)) %>% 
  rowwise() %>% 
  mutate(
    olap = length(intersect(
      tmp %>% filter(state == st1) %>% pull(gene),
      tmp %>% filter(state == st2) %>% pull(gene) )),
    pct = olap / length(unique(tmp %>% filter(state == st1) %>% pull(gene)))
  ) %>% ungroup()  %>% 
  mutate(st1 = factor(st1, levels = rev(c('early', 'mid','6','7')) ),
         st2 = factor(st2, levels = c('early', 'mid','6','7')))

```

```{r}
ggplot(olap, aes(x = st2, y = st1)) +
  geom_tile(aes(fill=pct), alpha = .8) + theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
  viridis::scale_fill_viridis("% Total", option = "C", direction = -1)+
  geom_text(aes(label = paste0(olap,'\n',signif(100*pct, 3), '%') ), size = 3) +
  labs(x='',y='', title = 'Transcriptomics Pseudotime State DE Genes') 
```

## combined
```{r}
rna <- bind_rows(rna.f.de %>% mutate(s = 'f'), 
                 bind_rows(rna.m.de, rna.m.de %>% filter(state == 6) %>% mutate(state = 7)) %>% 
                   mutate(s = 'm')) %>% 
    pivot_wider(id_cols = c(gene, state), 
                values_from = c(pvalue, effect), 
                names_from = s) %>% 
    rowwise() %>% 
    mutate(sig = case_when( 
        (pvalue_f <= 0.001 & pvalue_m > 0.001) ~ 'f_only',
        (pvalue_m <= 0.001 & pvalue_f > 0.001) ~ 'm_only',
        (pvalue_f > 0.001 & pvalue_m > 0.001) ~ 'neither',
        (pvalue_f <= 0.001 & pvalue_m <= 0.001) ~ 'both'),
        mn = mean(-log10(pvalue_f), -log10(pvalue_m), na.rm = T))
```

```{r fig.width=7, fig.height=5}
rna %>% 
  filter(sig != 'neither') %>% 
  ggplot(aes(effect_f, effect_m))+
  geom_hline(yintercept = 0, lwd =.1)+geom_vline(xintercept = 0, lwd =.1)+
  geom_point(aes(color = sig, size = mn), alpha = 0.3)+
  scale_size_continuous('mean -log10 pval', range = c(.1,2))+
  theme(legend.position = 'bottom')+
  facet_grid(sig~state)
```

```{r fig.width=9, fig.height=3}

cowplot::plot_grid(
  rna %>% 
    filter(sig != 'neither') %>% 
    group_by(state, sig) %>% 
    summarise(n_gene = length(unique(gene))) %>% 
    ggplot(aes(state, n_gene))+
    geom_col()+
    labs(subtitle = 'any effect direction')+
    facet_wrap(~sig),
  rna %>% 
    filter(sig == 'both', effect_f/effect_m > 0) %>% 
    group_by(state, sig) %>% 
    summarise(n_gene = length(unique(gene))) %>% 
    ggplot(aes(state, n_gene))+
    geom_col()+
    labs(subtitle = 'same effect direction')+
    facet_wrap(~sig)
  , rel_widths = c(1,.4)
)

```

```{r fig.width=9, fig.height=3}

tmp = rna %>% 
  filter (sig == 'both', effect_f/effect_m > 0 ) %>% 
  mutate( state = case_when(
    state %in% c(2,3) ~ 'early', 
    state %in% c(4,5) ~ 'mid', 
    T ~ as.character(state)) )


olap = crossing(st1 = unique(tmp$state), st2 = unique(tmp$state)) %>% 
  rowwise() %>% 
  mutate(
    olap = length(intersect(
      tmp %>% filter(state == st1) %>% pull(gene),
      tmp %>% filter(state == st2) %>% pull(gene) )),
    pct = olap / length(unique(tmp %>% filter(state == st1) %>% pull(gene)))
  ) %>% ungroup()  %>% 
  mutate(st1 = factor(st1, levels = rev(c('early', 'mid','6','7')) ),
         st2 = factor(st2, levels = c('early', 'mid','6','7')))

```

```{r}
ggplot(olap, aes(x = st2, y = st1)) +
  geom_tile(aes(fill=pct), alpha = .8) + theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
  viridis::scale_fill_viridis("% Total", option = "C", direction = -1)+
  geom_text(aes(label = paste0(olap,'\n',signif(100*pct, 3), '%') ), size = 3) +
  labs(x='',y='', title = 'Transcriptomics Pseudotime State DE Genes') 
```

# Networks
## early
## mid
## late
# Key Drivers

## early
```{r}
e0 <- read_tsv('/projects/carter-lab/caryg/network_tracing/results/pseudotime/prot_early/kda/prot_early_undirected_filt_node_edge_addWeights_edgeFactor_0.results.txt', col_types = cols())
e1 <- read_tsv('/projects/carter-lab/caryg/network_tracing/results/pseudotime/prot_early/kda/prot_early_undirected_filt_node_edge_addWeights_edgeFactor_1.results.txt', col_types = cols())
e1_top <- read_tsv('/projects/carter-lab/caryg/network_tracing/results/pseudotime/prot_early/kda/prot_early_undirected_filt_node_edge_addWeights_edgeFactor_1.tophits.txt', col_types = cols()) %>% 
  filter(FDR <= 0.05) %>% 
  mutate(kd = paste0(NODE,':',MODULE))
```

```{r fig.width=12, fig.height=10}
tmp <- bind_rows( e0 %>% mutate(ef = 'ef0_fdr'), e1 %>% mutate(ef = 'ef1_fdr')) %>%
  mutate(FDR = -log10(FDR)) %>%
  pivot_wider(id_cols = c(MODULE, NODE, MEMBER), names_from = ef, values_from = FDR) %>%
  mutate(across(contains('_fdr'), ~ if_else(is.na(.x), 0, .x))) %>% 
  mutate(delta_fdr = ef1_fdr - ef0_fdr) %>% arrange(desc(delta_fdr)) %>% 
  mutate(top_kd = if_else( paste0(NODE,':',MODULE) %in% e1_top$kd, 1, 0)) %>% 
  # rowwise() %>% 
  # mutate(MODULE = MODULE %>% str_split_fixed(., '_G:', 2) %>% .[1,2]) %>% 
  distinct() %>% 
  left_join(.,
            biodom %>% select(MODULE = GOterm_Name, Biodomain, abbr:color),
            by = 'MODULE')

early.enr.terms <- prot.f.enr %>% filter(state %in% 2:4, padj < 0.05) %>% pull(pathway)
mid.enr.terms <- prot.f.enr %>% filter(state %in% 5:6, padj < 0.05) %>% pull(pathway)

ggplot(tmp %>% filter(MODULE %in% early.enr.terms), 
       aes( ef1_fdr, delta_fdr ))+
  labs(x = 'wKDA FDR ef1, -log10', y = '\u0394 FDR, ef1 - ef0', 
       # subtitle = 'signif KDs (EF1 FDR < 0.05), \u0394 FDR > 0'
       subtitle = 'Early Pseudotime Branch NW KDs'
       )+
  geom_abline(intercept = 0, slope =1, lty= 2, lwd = .5)+
  geom_smooth(method = 'glm', lty = 2, lwd = .5, color = 'grey20')+
  geom_vline(xintercept = 0, lwd = .5)+ geom_hline(yintercept = 0, lwd = .5) +
  geom_point(alpha = .2, size = 1, aes(color = color) )+
  geom_point(data = subset(tmp, ef1_fdr > -log10(0.05) 
                           & top_kd == 1
                           & delta_fdr > 0
                           & MODULE %in% early.enr.terms
                           # & cell_filt == 'Micro'
                           )
             , alpha = .6, size = 2, aes(color = color))+
  ggrepel::geom_label_repel(
    data = subset( tmp, ef1_fdr > -log10(0.05)
                   & top_kd == 1
                   & MODULE %in% early.enr.terms
                   & delta_fdr > 0
                   # & cell_filt == 'Micro'
                   ),
    aes(label = paste0(NODE, '\n', MODULE)), #, '\n', Biodomain
    size = 2, alpha = .7, min.segment.length = 0)+
  facet_wrap(~Biodomain, scales = 'free')+ #
  scale_color_identity(guide = 'none') #; plotly::ggplotly(p)

# plotly::ggplotly(p2)
```
## mid

## late


# Enriched biodomains 

```{r fig.height=7, fig.width=10}

sig_level <- 1e-4

enr <- prot.f %>% 
  filter(!is.na(Biodomain)) %>% 
  mutate(leadingEdge = `enr2$leadingEdge` %>% str_remove_all(.,'^c|[:punct:]') %>% str_split(' ')  )

# count up the number of significant terms per biodomain
bdt <- bd.tally(enr$pathway[enr$padj <= sig_level], biodom) %>% 
  mutate(domain = fct_reorder(domain, n_sig_term, .desc = T)) %>% 
  arrange(domain)

# add biodomain annotations and plotting colors to enrichment results
enr <- enr %>% 
  mutate( Biodomain = fct_relevel(Biodomain, as.character(bdt$domain) )) %>% 
  mutate( state = fct_reorder( as.factor(state), as.numeric(state), .desc = T ))

# plot! -------------------------------------------------------------------
enr %>%   
  ggplot(aes(  NES, factor(state) ))+ 
  scale_color_identity()+ scale_fill_identity()+
  scale_size_continuous(
    limits = -log10(enr$padj) %>% range(),
    range = c(.7,5))+
  geom_jitter(
    data = subset(enr, padj > sig_level),
    color = 'grey85', alpha = .2, size = .5, show.legend = F
    )+
  geom_jitter(
    data = subset(enr, padj < sig_level ),
    aes(color = color, size = -log10(padj) ), #
    alpha = .5, show.legend = T
    )+
  geom_violin(
    data = subset(enr, padj < sig_level & NES < 0),
    scale='width', aes(fill = color), color = 'grey50',#, color = col
    alpha = .3, show.legend = F
    )+
  geom_violin(
    data = subset(enr, padj < sig_level & NES > 0),
    scale='width', aes(fill = color), color = 'grey50',#, color = col
    alpha = .3, show.legend = F
    )+
  geom_hline(yintercept = 0, lty = 3)+
  facet_wrap(~Biodomain)+
  labs(y='pseudotime state')+  theme(legend.position = 'right') +
  ggtitle(label = 'F proteomics pseudotime'
          # , subtitle = 'state pval < 0.05 / term padj < 0.05'
          , subtitle = paste0('term padj < ', sig_level)
          )

```

```{r fig.height=7, fig.width=10}

sig_level <- 1e-4

enr <- prot.m %>% 
  filter(!is.na(Biodomain)) %>% 
  mutate(leadingEdge = `enr2$leadingEdge` %>% str_remove_all(.,'^c|[:punct:]') %>% str_split(' ') )

# count up the number of significant terms per biodomain
bdt <- bd.tally(enr$pathway[enr$padj <= sig_level], biodom) %>% 
  mutate(domain = fct_reorder(domain, n_sig_term, .desc = T)) %>% 
  arrange(domain)

# add biodomain annotations and plotting colors to enrichment results
enr <- enr %>% 
  mutate( Biodomain = fct_relevel(Biodomain, as.character(bdt$domain) )) %>% 
  mutate( state = fct_reorder( as.factor(state), as.numeric(state), .desc = T ))

# plot! -------------------------------------------------------------------
enr %>%   
  ggplot(aes(  NES, factor(state) ))+ 
  scale_color_identity()+ scale_fill_identity()+
  scale_size_continuous(
    limits = -log10(enr$padj) %>% range(),
    range = c(.7,5))+
  geom_jitter(
    data = subset(enr, padj > sig_level),
    color = 'grey85', alpha = .2, size = .5, show.legend = F
    )+
  geom_jitter(
    data = subset(enr, padj < sig_level ),
    aes(color = color, size = -log10(padj) ), #
    alpha = .5, show.legend = T
    )+
  geom_violin(
    data = subset(enr, padj < sig_level & NES < 0),
    scale='width', aes(fill = color), color = 'grey50',#, color = col
    alpha = .3, show.legend = F
    )+
  geom_violin(
    data = subset(enr, padj < sig_level & NES > 0),
    scale='width', aes(fill = color), color = 'grey50',#, color = col
    alpha = .3, show.legend = F
    )+
  geom_hline(yintercept = 0, lty = 3)+
  facet_wrap(~Biodomain)+
  labs(y='pseudotime state')+  theme(legend.position = 'right') +
  ggtitle(label = 'M proteomics pseudotime'
          # , subtitle = 'state pval < 0.05 / term padj < 0.05'
          , subtitle = paste0('term padj < ', sig_level)
          )

```

```{r fig.height=7, fig.width=10}

sig_level <- 1e-6

enr <- r.rna.f %>% 
  filter(!is.na(Biodomain)
         # , abs(NES) > 2.5
         ) %>% 
  mutate(leadingEdge = `enr2$leadingEdge` %>% str_remove_all(.,'^c|[:punct:]') %>% str_split(' ') )

# count up the number of significant terms per biodomain
bdt <- bd.tally(enr$pathway[enr$padj <= sig_level], biodom) %>% 
  mutate(domain = fct_reorder(domain, n_sig_term, .desc = T)) %>% 
  arrange(domain)

# add biodomain annotations and plotting colors to enrichment results
enr <- enr %>% 
  mutate( Biodomain = fct_relevel(Biodomain, as.character(bdt$domain) )) %>% 
  mutate( state = fct_reorder( as.factor(state), as.numeric(state), .desc = T ))

# plot! -------------------------------------------------------------------
enr %>%   
  ggplot(aes(  NES, factor(state) ))+ 
  scale_color_identity()+ scale_fill_identity()+
  scale_size_continuous(
    limits = -log10(enr$padj) %>% range(),
    range = c(.7,5))+
  geom_jitter(
    data = subset(enr, padj > sig_level),
    color = 'grey85', alpha = .2, size = .5, show.legend = F
    )+
  geom_jitter(
    data = subset(enr, padj < sig_level ),
    aes(color = color, size = -log10(padj) ), #
    alpha = .5, show.legend = T
    )+
  geom_violin(
    data = subset(enr, padj < sig_level & NES < 0),
    scale='width', aes(fill = color), color = 'grey50',#, color = col
    alpha = .3, show.legend = F
    )+
  geom_violin(
    data = subset(enr, padj < sig_level & NES > 0),
    scale='width', aes(fill = color), color = 'grey50',#, color = col
    alpha = .3, show.legend = F
    )+
  geom_hline(yintercept = 0, lty = 3)+
  facet_wrap(~Biodomain)+
  labs(y='pseudotime state')+  theme(legend.position = 'right') +
  ggtitle(label = 'F transcriptomics pseudotime (AMP-AD v2.0)'
          # , subtitle = 'state pval < 0.05 / term padj < 0.05'
          , subtitle = paste0('term padj < ', sig_level)
          )

```

Something is funny about the new ROSMAP M pseudotime trajectories - all term directions seem inverted (e.g. Immune Response down, Synapse up) vs what might be expected and what is observed in other trajectories. Focus on F pseudotime trajectories for the moment.    
```{r fig.height=7, fig.width=10}

sig_level <- 1e-6

enr <- r.rna.m %>% 
  filter(!is.na(Biodomain)) %>% 
  mutate(leadingEdge = `enr2$leadingEdge` %>% str_remove_all(.,'^c|[:punct:]') %>% str_split(' ') )

# count up the number of significant terms per biodomain
bdt <- bd.tally(enr$pathway[enr$padj <= sig_level], biodom) %>% 
  mutate(domain = fct_reorder(domain, n_sig_term, .desc = T)) %>% 
  arrange(domain)

# add biodomain annotations and plotting colors to enrichment results
enr <- enr %>% 
  mutate( Biodomain = fct_relevel(Biodomain, as.character(bdt$domain) )) %>% 
  mutate( state = fct_reorder( as.factor(state), as.numeric(state), .desc = T ))

# plot! -------------------------------------------------------------------
enr %>%   
  ggplot(aes(  NES, factor(state) ))+ 
  scale_color_identity()+ scale_fill_identity()+
  scale_size_continuous(
    limits = -log10(enr$padj) %>% range(),
    range = c(.7,5))+
  geom_jitter(
    data = subset(enr, padj > sig_level),
    color = 'grey85', alpha = .2, size = .5, show.legend = F
    )+
  geom_jitter(
    data = subset(enr, padj < sig_level ),
    aes(color = color, size = -log10(padj) ), #
    alpha = .5, show.legend = T
    )+
  geom_violin(
    data = subset(enr, padj < sig_level & NES < 0),
    scale='width', aes(fill = color), color = 'grey50',#, color = col
    alpha = .3, show.legend = F
    )+
  geom_violin(
    data = subset(enr, padj < sig_level & NES > 0),
    scale='width', aes(fill = color), color = 'grey50',#, color = col
    alpha = .3, show.legend = F
    )+
  geom_hline(yintercept = 0, lty = 3)+
  facet_wrap(~Biodomain)+
  labs(y='pseudotime state')+  theme(legend.position = 'right') +
  ggtitle(label = 'M transcriptomics pseudotime (AMP-AD v2.0)' 
          # , subtitle = 'state pval < 0.05 / term padj < 0.05'
          , subtitle = paste0('term padj < ', sig_level)
          )

```

# Dump input gene lists 

```{r}

enr <- prot.f %>% 
  filter(!is.na(Biodomain), padj < 1e-4) %>% 
  mutate(leadingEdge = `enr2$leadingEdge` %>% str_remove_all(.,'^c|[:punct:]|\\n') %>% str_split(' ') ) %>% 
  group_by( state ) %>% 
  summarise( leadingEdge_genes = leadingEdge %>% unlist() %>% unique() %>% list() ) %>% 
  unnest(leadingEdge_genes)

for(s in unique(enr$state)){ 
  enr %>%
    filter(state == s) %>%
    select(leadingEdge_genes) %>%
    write_tsv(
      .,
      paste0( here::here(), '/results/input_gene_lists/proteomics_f_pseudotime_state',s,'_leadingEdge.tsv'),
      col_names = F)
  }

```

```{r}

enr <- r.rna.f %>% 
  filter( !is.na(Biodomain), padj < 1e-6 ) %>% 
  mutate(leadingEdge = `enr2$leadingEdge` %>% str_remove_all(.,'^c|[:punct:]|\\n') %>% str_split(' ') ) %>% 
  group_by( state ) %>% 
  summarise( leadingEdge_genes = leadingEdge %>% unlist() %>% unique() %>% list() ) %>% 
  unnest(leadingEdge_genes)

for(s in unique(enr$state)){ 
  enr %>%
    filter(state == s) %>%
    select(leadingEdge_genes) %>%
    write_tsv(
      .,
      paste0( here::here(), '/results/input_gene_lists/transcriptomics_f_pseudotime_state',s,'_leadingEdge.tsv'),
      col_names = F)
  }

```


<!-- # TODO: Submit path trace slurm jobs -->

<!-- ```{r} -->
<!-- f = list.files(paste0(here::here(), '/results/input_gene_lists'), full.names = T) -->

<!-- # cmd = paste0( -->
<!-- #   'module load singularity; sbatch ', here::here(), '/scripts/batch_trace.sbatch ', f[1] -->
<!-- #   ) -->

<!-- Sys.setenv(cmd = paste0( -->
<!--   'sbatch ', here::here(), '/scripts/batch_trace.sbatch ', f[1] -->
<!--   )) -->
<!-- ``` -->

<!-- ```{r engine='bash'} -->

<!-- $cmd -->

<!-- ``` -->

# Analyze traced NWs
```{r}

# load base NW
load( synapser::synGet('syn22992753')$path )
baseNW = net_undirected
rm(net_undirected)

# what were the inputs to the trace
inputs <- map_dfr( 
  unique(prot.f$state), 
  ~prot.f %>% 
    filter(!is.na(Biodomain), padj < 1e-4, state == .x) %>% 
    mutate(leadingEdge = `enr2$leadingEdge` %>% str_remove_all(.,'^c|[:punct:]|\\n') %>% str_split(' ') ) %>% 
    group_by( Biodomain, pathway, color ) %>% 
    summarise( leadingEdge_genes = leadingEdge %>% unlist() %>% unique() %>% list() ) %>% 
    unnest(leadingEdge_genes) %>% 
    mutate(state = .x)
) %>% ungroup()


```

```{r}

nws = tibble( state = unique(inputs$state),
              n_input = map_dbl(unique(inputs$state), 
                                ~ inputs %>% filter(state == .x) %>% select( leadingEdge_genes ) %>% distinct() %>% nrow() ),
              induced_V = map_dbl(unique(inputs$state), 
                                  ~ induced_subgraph( baseNW, 
                                                      v = which( (V(baseNW) %>% names()) %in% 
                                                                   (inputs %>% filter(state == .x) %>% 
                                                                      pull(leadingEdge_genes)) ) ) %>% 
                                    V() %>% length() ),
              induced_E = map_dbl(unique(inputs$state), 
                                  ~ induced_subgraph( baseNW, 
                                                      v = which( (V(baseNW) %>% names()) %in% 
                                                                   (inputs %>% filter(state == .x) %>% 
                                                                      pull(leadingEdge_genes)) ) ) %>% 
                                    E() %>% length() )
              )


# list traced NW objects
nw.files = list.files( paste0(here::here(), '/results/NW/pseudotime/'), full.names = T ) %>% 
  str_subset(., '_undirected')

# read NW objects into a list
nw <- map( nw.files, ~ read_graph( .x, format = 'graphml') )

nws$trace_undir_V = map_dbl( 1:length(nw), ~ V(nw[[.x]]) %>% length() )
nws$trace_undir_E = map_dbl( 1:length(nw), ~ E(nw[[.x]]) %>% length() )

# list traced NW objects
nw.files = list.files( paste0(here::here(), '/results/NW/pseudotime/'), full.names = T ) %>% 
  str_subset(., '_directed')

# read NW objects into a list
nw <- map( nw.files, ~ read_graph( .x, format = 'graphml') )

nws$trace_dir_V = map_dbl( 1:length(nw), ~ V(nw[[.x]]) %>% length() )
nws$trace_dir_E = map_dbl( 1:length(nw), ~ E(nw[[.x]]) %>% length() )

```

```{r}
DT::datatable(nws, rownames = F)
```

As a fraction of the total nodes and edges present in the base network
```{r}
nws %>% 
  mutate( across( .cols = contains('V'), .fns = ~ .x / base_nV ), 
          across( .cols = contains('E', ignore.case = F), .fns = ~ .x / base_nE ) ) %>% 
  DT::datatable(., rownames = F) %>% 
  DT::formatSignif(., columns = names(nws) %>% str_subset(., '_V|_E') )
```

# Upload to synapse
```{r}
synLogin()

nw.files = list.files( paste0(here::here(), '/results/NW/pseudotime'), full.names = T ) 
id = 'syn50864398'

for( f in nw.files ){
  n <- str_extract(f, 'state.*graphml')
  foo <- synStore( File(f, parent = id, name = n) )
}

```

# Nanostring panel genes NW
```{r}
ns = readxl::read_xls('/projects/carter-lab/caryg/datamouse/data/Supplemental_Table_3_Nanostring_Gene_Annotations.xls') %>% 
  select(gene = `Human Gene Symbol`) %>% distinct() 

nsNW = read.graph('results/NW/nanostring_traced_subgraph.graphml', 
                  format = 'graphml')

ns_inducedNW = induced_subgraph(
  baseNW, 
  vids = which( (V(baseNW) %>% names()) %in% ns$gene ) 
  )
write.graph(ns_inducedNW, 'results/NW/nanostring_induced_subgraph.graphml', format = 'graphml')

nw.files = list.files( paste0(here::here(), '/results/NW'), 
                        pattern = 'nano', full.names = T ) 
id = 'syn50864711'

for( f in nw.files ){
  n <- str_extract(f, 'nanostring.*graphml')
  foo <- synStore( File(f, parent = id, name = n) )
}

```

